TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/ReplayEngine.java

```diff
@@ -8,58 +9,92 @@
 import java.util.function.BiConsumer;
 import java.util.function.Consumer;
 
+@Slf4j
 public class ReplayEngine implements BiConsumer<String, TrafficObservation> {
-    private final Map<String, RequestResponsePacketPair> liveStreams;
+    enum State {
+        NOTHING_SENT,
+        REQUEST_SENT
+    }
+    public static class Accumulation {
+        RequestResponsePacketPair rrPair = new RequestResponsePacketPair();
+        State state = State.NOTHING_SENT;
+    }
+    private final Map<String, Accumulation> liveStreams;
+    private final Consumer<HttpMessageAndTimestamp> requestHandler;
     private final Consumer<RequestResponsePacketPair> fullDataHandler;
 
-    public ReplayEngine(Consumer<RequestResponsePacketPair> fullDataHandler) {
+    public ReplayEngine(Consumer<HttpMessageAndTimestamp> requestReceivedHandler,
+                        Consumer<RequestResponsePacketPair> fullDataHandler) {
         liveStreams = new HashMap<>();
+        this.requestHandler = requestReceivedHandler;
         this.fullDataHandler = fullDataHandler;
     }
 
     @Override
     public void accept(String id, TrafficObservation observation) {
-        boolean updateFirstResponseTimestamp = false;
-        RequestResponsePacketPair runningList = null;
         var pbts = observation.getTs();
         var timestamp = Instant.ofEpochSecond(pbts.getSeconds(), pbts.getNanos());
         if (observation.hasEndOfMessageIndicator()) {
-            publishAndClear(id);
+            handleEndOfMessage(id, getAccumulationForFirstRequestObservation(id));
```

Comment:
Looks weird that we could potentially call handleEndOfMessage 2x here

Reply:
Agree, but this also goes away in the next PR